### PR TITLE
Mock Array Big - Power Grid and macro placement modification

### PR DESCRIPTION
**Pull request by louiic**
Updating Power Grid strategy for Mock-Array and Element to includes power ring around the design.

Modify / correct macro placement
    - correct macro size and core margin to align with placement and routing track.
    - increase power grid density
 
Modify / correct mock-array
    - add power ring
    - reduce margin around core

Modify / correct macro placement script
    - taking user's row and column input
    - taking user's Element placement pitch

### CODE REVIEW

File: .gitmodules [hunk 1]
```diff
@@ -3,7 +3,7 @@
 	url = https://github.com/The-OpenROAD-Project/yosys.git
 [submodule "tools/OpenROAD"]
 	path = tools/OpenROAD
-	url = ../OpenROAD.git
+	url = https://github.com/The-OpenROAD-Project/OpenROAD.git
```
Comment (tspyrou) [inline .gitmodules]: @louiic is this change really needed?

**Review by tspyrou — CHANGES_REQUESTED** [review]

**Comment by oharboe** [comment]
@tspyrou Can you add @yupferris as a reviewer?

**Comment by maliberty** [comment]
You'll need to update the metrics with update_ok:
```
[ERROR] placeopt__design__instance__area fail test: 11118.0 <= 10996.0
[ERROR] finish__design__instance__area fail test: 11119.6 <= 11004.0
```

File: .gitmodules (hunk 1 above)
Comment (louiic) [reply .gitmodules]: Yes ... without this, the fork can never find OpenROAD GIT

Comment (tspyrou) [reply .gitmodules]: @louiic It took me a little while to remember why this was important. The reason it was done with ../ is so that a clone from the -private project would clone the OpenROAD submodule from the -private project. Also clones from forks will clone the forked openroad submodule. In your case, if you made a fork of openroad as well as openroad-flow-scripts you should be able to run without this change. The ../ is not strictly required but for the users of forks and -private its confusing when ORFS comes from their fork but OR does not.  Its probably good for your OR submodule if you turn on the action to sync from its parent so then there is no diff between the two.

**Review by tspyrou — CHANGES_REQUESTED** [review]

File: flow/designs/asap7/mock-array-big/Element/config.mk [hunk 2]
```diff
@@ -16,3 +16,5 @@ export DIE_AREA = $(shell python3 designs/asap7/mock-array-big/ce_die_area.py)
 export IO_CONSTRAINTS = designs/asap7/mock-array-big/Element/io.tcl
 
 export PDN_TCL = designs/asap7/mock-array-big/Element/pdn.tcl
+
+export MAX_ROUTING_LAYER = M5
```
Comment (oharboe) [inline flow/designs/asap7/mock-array-big/Element/config.mk:22]: add a comment above as to why this is done? Is this required to get PDN rings to work or is it an additional improvement?

File: flow/designs/asap7/mock-array-big/Element/pdn.tcl [hunk 3]
```diff
@@ -3,17 +3,24 @@
...
 set_voltage_domain -name {CORE} -power {VDD} -ground {VSS}
+
 ####################################
 # standard cell grid
 ####################################
 define_pdn_grid -name {top} -voltage_domains {CORE}
-add_pdn_stripe -grid {top} -layer {M1} -width {0.018} -pitch {0.54} -offset {0} -followpins
-add_pdn_stripe -grid {top} -layer {M2} -width {0.018} -pitch {0.54} -offset {0} -followpins
-add_pdn_stripe -grid {top} -layer {M5} -width {0.12} -spacing {0.072} -pitch {11.88} -offset {0.300}
+add_pdn_ring    -grid {top} -layers {M5 M4} -widths {0.12 0.12} -spacings {0.072} -core_offset {0.084}
```
Comment (oharboe) [inline flow/designs/asap7/mock-array-big/Element/pdn.tcl:21]: add some comments here as to how these magic numbers where chosen and what requirements they satisfy.

File: flow/designs/asap7/mock-array-big/ce_die_area.py [hunk 4]
```diff
@@ -1,3 +1,6 @@
 import config
 
-print(f'0 0 {config.pitch} {config.pitch}')
+margin_x = config.placement_grid_x * 8
```
Comment (oharboe) [inline flow/designs/asap7/mock-array-big/ce_die_area.py:3]: delete the two unused lines (margin_x and margin_y)

File: flow/designs/asap7/mock-array-big/config.py [hunk 5]
```diff
@@ -1,27 +1,44 @@
-# see flow/platforms/asap7/openRoad/pdn/grid_strategy-M2-M5-M7.tcl
 import os
 
+# number of Elements in row and column, user can set via environment variable
 rows = int(os.environ.get("MOCK_ARRAY_HEIGHT", "8"))
 cols = int(os.environ.get("MOCK_ARRAY_WIDTH", "8"))
+
+# must be 8x8
+if (rows * cols) < 64:
+  print("ERROR: number of elements row and columns must add up to 64.")
```
Comment (oharboe) [inline flow/designs/asap7/mock-array-big/config.py]: I would like to avoid this restriction, can it simply be removed?

We are investigating other layouts, like a really large element with 1x1 array or a large 16x8 array.

File: flow/designs/asap7/mock-array-big/config.py [hunk 6]
```diff
@@ -1,27 +1,44 @@
...
-pitch_and_margin = pitch + margin
-
-core_width = cols * pitch_and_margin + core_offset_x
-core_height = rows * pitch_and_margin + core_offset_y
-die_width = core_width + margin * 2
-die_height = core_height + margin * 2
+# must be > 1
+if pitch_scale < 1:
+  print("ERROR: Element placement pitch must be greater than 1")
+  exit
```
Comment (oharboe) [inline flow/designs/asap7/mock-array-big/config.py]: `exit` is not python. Remove print() line and simply add `raise Exception("Element placement pitch must be greater than 1")

File: flow/designs/asap7/mock-array-big/config.py [hunk 7]
```diff
@@ -1,27 +1,44 @@
...
-die_offset_x = core_offset_x + margin
-die_offset_y = core_offset_y + margin
-
-pitch_and_margin = pitch + margin
-
-core_width = cols * pitch_and_margin + core_offset_x
-core_height = rows * pitch_and_margin + core_offset_y
-die_width = core_width + margin * 2
-die_height = core_height + margin * 2
+# must be > 1
```
Comment (oharboe) [inline flow/designs/asap7/mock-array-big/config.py]: Delete superfluous comment

File: flow/designs/asap7/mock-array-big/die_area.py [hunk 8]
```diff
@@ -1,3 +1,4 @@
 import config
 
+#print(f'0 0 {config.core_width + (2 * config.margin_x)} {config.core_height + (2 * config.margin_y)}')
```
Comment (oharboe) [inline flow/designs/asap7/mock-array-big/die_area.py]: delete commented out line

File: flow/designs/asap7/mock-array-big/macro-placement.tcl [hunk 9]
```diff
@@ -1,18 +1,38 @@
 # fish out values from single source of truth: config.py
-set vals [regexp -all -inline {\S+} [exec sh -c {cd designs/asap7/mock-array-big && python3 -c "import config;print(f'{config.rows} {config.cols} {config.pitch_and_margin} {config.die_offset_x} {config.die_offset_y}')"}]]
-lassign $vals rows cols pitch_and_margin die_offset_x die_offset_y
+set vals [regexp -all -inline {\S+} [exec sh -c {cd designs/asap7/mock-array-big && python3 -c "import config;print(f'{config.rows} {config.cols} {config.ce_pitch_x} {config.ce_pitch_y} {config.margin_x} {config.margin_y} {config.placement_grid_x} {config.placement_grid_y}')"}]]
```
Comment (oharboe) [inline flow/designs/asap7/mock-array-big/macro-placement.tcl:2]: Bless you for figuring out this nasty bit of code :-)

Ideally I could simply stick to Python across the board, but Python snippets can't be invoked from OpenROAD when OpenROAD is invoked with a .tcl script.

File: flow/designs/asap7/mock-array-big/macro-placement.tcl [hunk 10]
```diff
@@ -1,18 +1,38 @@
...
+lassign $vals rows cols ce_pitch_x ce_pitch_y margin_x margin_y placement_grid_x placement_grid_y
 
 set block [ord::get_db_block]
 set units [$block getDefUnits]
 
+set x [expr ($margin_x + ($ce_pitch_x / 2)) * $units]
+set y [expr ($margin_y + ($ce_pitch_y / 2)) * $units]
+
+set instList [list]
+for {set i 0} {$i < 8} {incr i} {
```
Comment (oharboe) [inline flow/designs/asap7/mock-array-big/macro-placement.tcl]: This code should be reverted, such that other configurations like 1x1, 16x8 can be supported.

I can't surmise why it was changed to be specific to 8x8 configurations.

**Review by oharboe — CHANGES_REQUESTED** [review]
Add @yupferris  as a reviewer, he knows something about PDNs and is the one who will have to apply this result.

**Review by yupferris — APPROVED** [review]
Looks good to me, provided that the comments from @oharboe are addressed. Great stuff!

File: flow/designs/asap7/mock-array-big/macro-placement.tcl [hunk 11]
```diff
@@ -1,18 +1,23 @@
...
 
+set x [expr ($margin_x + ($ce_pitch_x / 2)) * $units]
+set y [expr ($margin_y + ($ce_pitch_y / 2)) * $units]
+
 for {set row 0} {$row < $rows} {incr row} {
   for {set col 0} {$col < $cols} {incr col} {
     set inst [$block findInst [format "ces_%d_%d" $row $col]]
     $inst setOrient R0
-    set x [expr round([expr {$die_offset_x + $col * $pitch_and_margin}] * $units)]
-    set y [expr round([expr {$die_offset_y + $row * $pitch_and_margin}] * $units)]
```
Comment (oharboe) [inline flow/designs/asap7/mock-array-big/macro-placement.tcl:13]: nit: rather than having mutable iteration variables, I would simply calculate the expression here as before in a single place. x/y calcs are now spread over 5 lines instead of two lines close to their use.

**Review by oharboe — CHANGES_REQUESTED** [review]
@maliberty I suggest rebase on master and squash this pull request for further review and to simplify the git history on master.

**Review by tspyrou — APPROVED** [review]
